### ISSUE TITLE
Crawling using cannonical URLs

### ISSUE THREAD

**Issue by leonardsaers** [Norconex/collector-http#79]
Norconex collector-http is a very competent crawler. There is now only one feature which I'm missing and that's crawling using canonical links:
http://googlewebmastercentral.blogspot.se/2009/02/specify-your-canonical.html

This might be hard to implement. Nutch have had this issue opened for the last 5 years.
https://issues.apache.org/jira/browse/NUTCH-710

I have several times encountered different CMSs which has several URLs pointing to the same page, where there may be small changes in the HTML depending on which URL has been used. This behaviour can normally not be changed, but the CMS can normally provide a canonical URL.

When using a crawler which doesn’t handles canonical links for those situations is hard since the crawl database will be flooded with URLs that are actually pointing to the same page. A CMS which contains 1 000 site might be represented by 1 000 000 URLs in the crawl database. This makes crawling really slow. 

It would be nice if Norconex could group a set of URLs to one canonical URL.

**Comment by essiembre** [comment]
Thanks for the good words on the crawler! That's our reward :-)

I'd be more than happy to add such a nice feature for sure (hopefully within the next 5 years ;-) ).  It should not be TOO difficult to swap a URL with its canonical form when known... Do we have reliable way to always find those besides the ``<link rel="canonical" href="..." />`` tag?   Leveraging that would already be a great start, but do you know of any other reliable indicators?

The reason I did not think this was a big issue before is because of the content checksummer feature that will ensure uniqueness when committing documents.  

This being said, having to crawl 100 URL variations like you suggest, that all need to be crawled/downloaded just to figure out they are the same can really slow the crawl down though, you are right.  BUT: how else can we figure out a URL canonical form BEFORE downloading its content and checking the ``<link />`` tag?   Unless the pattern is known somewhere ahead of time, isn't downloading each file the only way? 

Because if a pattern is known upfront, there is already the [GenericURLNormalizer](http://www.norconex.com/collectors/collector-http/latest/apidocs/com/norconex/collector/http/url/impl/GenericURLNormalizer.html) class that can normalize URLs to eliminate semantically identical ones before they are downloaded (lightning fast).

I would like to hear your thoughts on this.

**Comment by essiembre** [comment]
After reading the links you provided, I see Nutch is thinking of simply rejecting pages with the canonical href in them.  If we ensure the referenced URL gets indexed properly, do you see any issues with this approach?  We would have to ensure that if the canonical URL specified is the same as the current URL we don't reject that one.   

With such an approach, we would still need to download the page but at least we would not have to perform a checksum on its content and we could potentially skip some other steps in the import flow.  So we would save a bit of processing, but the the initial download cost would still be there.

I wonder if the best approach isn't to encourage people to detect and apply good normalization patterns for sites they know well to help prevent unnecessary downloads (the GenericURLNormalizer allowing regex-based search and replace).   Not sure how much detection/automation we can do here to prevent downloads.

**Comment by leonardsaers** [comment]
I would say that norconex should only consider <link rel="canonical" href="..." /> to be meta data for canonical URLs.

Pattern will work in many cases, but they need to be maintained. The CMS can be upgraded which may results in new patterns to consider. Some time, it's just hard finding a pattern at all.

I would try to solve this with as simple approach as possible. The main issue is re-crawling of same content when several URLs is pointing to one canonical URL. Each discovered URL must be downloaded once in order to figure out if it has a canonical URL. But the crawler should try to only crawl a page one time, and it need to remember which URL points to a canonical URL in order to reduce the times a page is being fetched.

Information about URLs pointing to a canonical URL can be used to reject an URL from being fetched during re-crawl. This would result in faster re-crawls. Maybe this could be solved by a bloom filter (http://en.wikipedia.org/wiki/Bloom_filter), where URLs pointing to a canonical URL is added to the bloom filter. That would reduce the risk of filling up the crawl data base with non canonical URLs.

Just rejecting non canonical URLs would work, but the crawler would still need to re- fetch the non canonical links when doing a re-crawl.

**Comment by essiembre** [comment]
Just highlighting what might already be obvious to you: within any given crawl session, each distinct URLs can be downloaded zero or one time, never more.  So if many pages are pointing to the same canonical URL, that target URL will be downloaded only once regardless.

So it is the many URL variations that we are concerned about and I really like your idea of flagging the non canonical URLs so on subsequent crawls, they are not downloaded again and simply ignored.  That way only the initial crawl will be more costly. 

This raises an important concern though: what if the pages behind a non-canonical URL changes?  What if it drops its ``<link rel="canonical" href="..." />`` tag and becomes a legitimate distinct page? That change will never be picked up.  Unlikely scenario, but certainly possible.

I think we could make that non-canonical flagging feature optional, but I wonder if maybe we should allow for some kind of flag that says to recheck those flagged non-canonical URLs after X amount of time, in case the unlikely scenario I mention occurs.  What do you think?  Would it cover all angles?

**Comment by leonardsaers** [comment]
Yes, I guess that the initial crawl have to be costly. It would be preferable if the crawler only would download the canonical URL and not the URLs pointing to it. But new URLs needs to be downloaded before the canonical URL data can be parsed. As a distinct URL could sort of be downloaded several times if all variants of URLs pointing to a canonical URLs counts as the one distinct canonical URL.

I don't think norconex needs to handle the problem if the page behind the canonical link change. It will result in a page which is not being crawled and it's important that the person setting up the crawler knows about this behaviour. Instead, as you suggest, crawling using canonical link should be optional. Maybe the default should be to disable the feature since there might be a risk of a page not being crawled if first pointing to a canonical URL and later becoming a canonical page.

**Comment by essiembre** [comment]
Thanks for your great feedback.  This feature will definitely be implemented.

**Comment by essiembre** [comment]
I implemented a solution for this.  Starting with 2.2.0, the HTTP Collector will check for canonical URLs defined in both HTTP headers and in HTML pages. Documents having one will be rejected in favor of the canonical one.  This feature is turned on by default.  

A new interface was created for this feature and a default implementation called [GenericCanonicalLinkDetector](http://www.norconex.com/collectors/collector-http/latest/apidocs/com/norconex/collector/http/url/impl/GenericCanonicalLinkDetector.html) was created.  Disabling it can be done quickly in a crawler section of your config:

```xml
 <canonicalLinkDetector ignore="true" />
```
Note that skipping page with canonical URLs on incremental crawls was not implemented. After some thoughts, it turns out sites that specify a canonical URL in their HTTP Headers will prevent a document from being downloaded, even on first crawl.  For the HTML pages with the canonical ``<link>`` in their ``<head>``, skipping them on incremental run can be accomplished by relying on the HTTP headers checksum (the last modified date by default). If the page has not changed, it won't be downloaded again.  That should cover the vast majority of scenarios, and I think it is safer to download and check it again in the event that it changed, in case the canonical reference was removed or got changed.

The latest [snapshot release](http://www.norconex.com/collectors/collector-http/download) has this new feature.
